rfcs/20201121-keras-model-fit-ps.md

```diff
@@ -0,0 +1,501 @@
+
+This problem can be solved by a new `DataHandler` class which has customized logic to configure the datasets and steps. In addition, it provides how synchronization with the cluster is done, and how to interpret the result from the train function.
+
+
+```
+class ClusterCoordinatorDataHandler(DataHandler):
+  """A `DataHandler` that is compatible with `ClusterCoordinator`."""
+
+  def _configure_dataset_and_inferred_steps(self, strategy, x, steps_per_epoch,
+                                            class_weight):
+    self._dataset = self._model._cluster_coordinator.create_per_worker_dataset(x)
+    if steps_per_epoch is None:
+      raise RuntimeError(
+          "Steps per epoch must be specified with `ParameterServerStrategy`.")
+    self._inferred_steps = steps_per_epoch
+
+  def sync(self):
+    self._model._cluster_coordinator.join()
+
+  def resolve_logs(self, logs):
+    return logs.fetch()
+```
+
+
+And, in the existing `DataHandler`,
+
+
+```
+class DataHandler(object):
+
+  def _configure_dataset_and_inferred_steps(self, strategy, x, steps_per_epoch,
+                                            class_weight):
+    # Does what it does currently
+
+  def sync(self):
+    context.async_wait()
+
+  @contextlib.contextmanager
+  def catch_stop_iteration(self):
+    """Catches errors when an iterator runs out of data."""
+    try:
+      yield
+      self.sync()
+    except (StopIteration, errors.OutOfRangeError):
+      # stop the iteration
+
+  def resolve_logs(self, logs):
+    return logs
+```
+
+
+A new `DataAdapter` is also needed to make sure the training API knows a dataset factory is a supported path:
+
+
+```
+class DataFactoryAdapter(DataAdapter):
+  """Adapter that handles dataset functions."""
+
+  @staticmethod
+  def can_handle(x, y=None):
+    if isinstance(x, DatasetFactory):
+      assert y is None
+      return True
+    return False
+```
+
+
+
+#### Callbacks
+
+With `ParameterServerStrategy`, the return value of `Model.train_function` is a dict `RemoteValue`s. This dict is passed as the `logs` argument to the `CallbackList` object. To obtain the `NumPy` value of each `RemoteValue`, one can do:
+
+
+```
+def to_numpy(logs):
+  cluster_coordinator.join()  # Sync the workers.
+  return {k: v.fetch() for k, v in logs.items()}  # Return the NumPy results
+```
+
+
+This logic can be handled by the `CallbackList` object, which already handles converting `tf.Tensor` logs to `NumPy`. However, obtaining these logs requires us to sync the workers, which will result in a slowdown if done every batch. Currently we only plan to sync the workers once every epoch at the end of the epoch. 
```

Comment:
We should not have callbacks include PS-specific logic. The task of synchronizing (and deciding when it's appropriate) should be done in the Model.

Note that it is necessary for us to provide a way to synchronize and run callbacks at higher frequency than once per epoch. This is critical. We could use `steps_per_execution` to configure this (with a high default value for the PS use case).

Reply:
It is desirable that some callbacks like checkpointing callbacks will be invoked in the middle of an epoch. But increasing the `steps_per_execution` to a large value has disadvantages: these steps inside a single function call cannot be distributed across workers for load-balancing purposes; worker failure (leading to the interruption of a function execution) will retry all these `steps_per_execution` steps.

Also checkpointing doesn't necessary need synchronization in asynchronous training. IMO it would be useful for Keras to support time-based callback for checkpointing that doesn't need synchronization. The checkpointing, especially in the middle of an epoch, is mainly for failure recovery purposes and therefore time-based callbacks would easily allow users to know that at most X minutes of training would be lost if there is PS failure.

Reply:
> We should not have callbacks include PS-specific logic

Agree, really the only change needed is in `tf_utils.to_numpy_or_python_type`, which is used by `CallbackList` when it detects that it needs to convert `Tensor` logs into `NumPy` for a batch. A user would write Callbacks without having to known anything about this, they'll still see NumPy data.

The change would roughly be:

```
def to_numpy_or_python_type(data):
  # Sync workers and get NumPy values
  if isinstance(data, RemoteValue):
    get_strategy().join()
    return data.fetch()
  else:
    # Current logic stays the same for other strategies.
    ...
```

Reply:
Re: The task of synchronizing (and deciding when it's appropriate) should be done in the Model, one option is having `logs` resolved in `model.fit`, after the steps are done, and before epoch_end callbacks are called. Concretely, in the current code, we do:

```
      for epoch, iterator in data_handler.enumerate_epochs():
        ...
        with data_handler.catch_stop_iteration():  # This will sync
          for step in data_handler.steps():
              tmp_logs = self.train_function(iterator)
              logs = tmp_logs
              ...
        logs = data_handler.resolve_logs(logs)  # convert remote to concrete
        if logs is None:
          raise ValueError('Expect x to be a non-empty array or dataset.')
        epoch_logs = copy.copy(logs)

        # Run validation.
        ...

        callbacks.on_epoch_end(epoch, epoch_logs)  # epoch_logs are concrete values
```

With this, we would just need to provide an implementation for `resolve_logs`. In the case with `ClusterCoordinatorDataHandler`, it would be `logs.fetch()`, whereas the default is an no-op. I believe this could avoid a change in `tf_utils.to_numpy_or_python_type`.

Reply:
I think it's OK to have the logic in `tf_utils.to_numpy_or_python_type` (this utility is basically already just "the thing that takes values returned by Model methods and converts them to what the user should see"

IIUC, Francois's concern was that (please correct me if I'm wrong Francois) users who write `Callback`s shouldn't have to be aware of any of this, shouldn't have to know what a `RemoteValue` is, etc. 

As long as we achieve that, either `data_handler.resolve_logs` or `tf_utils.to_numpy_or_python_type` seems fine to me, with a slight preference for `tf_utils` because it consolidates all of our `logs` munging into one place

Reply:
Sounds good Tom - and thanks for updating the callback section which reflects the current consensus.